docs/dev/core/internal/upload.md

```diff
@@ -0,0 +1,69 @@
+# Upload mechanism
+
+The `glean-core` Rust crate does not handle the ping upload directly.
+Network stacks vastly differ between platforms, applications and operating systems.
+It is best to use the available platform capabilities to implement any network communication.
+
+Glean still controls all upload and coordinates the platform side with its own internals.
+All wrappers implement ping uploading around a common API.
+
+## Upload task API
+
+The following diagram visualizes the communication between Glean core (the Rust crate),
+a Glean wrapper (e.g. the Kotlin or Swift implementation) and a Glean end point server.
+
+```mermaid
+sequenceDiagram
+    participant Glean core
+    participant Glean wrapper
+    participant Server
+
+    Glean wrapper->>Glean core: get_upload_task()
+    Glean core->>Glean wrapper: Task::Upload(task)
+    Glean wrapper-->>Server: POST /submit/{task.id}
+    Server-->>Glean wrapper: 200 OK
+    Glean wrapper->>Glean core: upload_response(200)
+    Glean wrapper->>Glean core: get_upload_task()
+    Glean core->>Glean wrapper: Task::Done
+```
+
+Glean core will take care of file management, cleanup, rescheduling and throttling[^1].
+
+> [^1] Note: At this point throttling is not implemented. Follow [Bug 1543612](https://bugzilla.mozilla.org/show_bug.cgi?id=1543612) for updates.
+
+## Available APIs
+
+### Rust
+
+For direct Rust consumers the global `Glean` object provides these methods:
+
+```rust
+/// Gets the next task for an uploader. Which can be either:
```

Comment:
Which can be either?

Reply:
oh. yeah, I should have not copied the old stuff.

Reply:
```suggestion
/// Gets the next task for an uploader.
```